source/common/network/socket_option_factory.cc

```diff
@@ -33,27 +33,55 @@ SocketOptionFactory::buildTcpKeepaliveOptions(Network::TcpKeepaliveConfig keepal
 std::unique_ptr<Socket::Options> SocketOptionFactory::buildIpFreebindOptions() {
   std::unique_ptr<Socket::Options> options = absl::make_unique<Socket::Options>();
   options->push_back(std::make_shared<Network::AddrFamilyAwareSocketOptionImpl>(
-      Network::Socket::SocketState::PreBind, ENVOY_SOCKET_IP_FREEBIND, ENVOY_SOCKET_IPV6_FREEBIND,
-      1));
+      envoy::api::v2::core::SocketOption::STATE_PREBIND, ENVOY_SOCKET_IP_FREEBIND,
+      ENVOY_SOCKET_IPV6_FREEBIND, 1));
   return options;
 }
 
 std::unique_ptr<Socket::Options> SocketOptionFactory::buildIpTransparentOptions() {
   std::unique_ptr<Socket::Options> options = absl::make_unique<Socket::Options>();
   options->push_back(std::make_shared<Network::AddrFamilyAwareSocketOptionImpl>(
-      Network::Socket::SocketState::PreBind, ENVOY_SOCKET_IP_TRANSPARENT,
+      envoy::api::v2::core::SocketOption::STATE_PREBIND, ENVOY_SOCKET_IP_TRANSPARENT,
       ENVOY_SOCKET_IPV6_TRANSPARENT, 1));
   options->push_back(std::make_shared<Network::AddrFamilyAwareSocketOptionImpl>(
-      Network::Socket::SocketState::PostBind, ENVOY_SOCKET_IP_TRANSPARENT,
+      envoy::api::v2::core::SocketOption::STATE_BOUND, ENVOY_SOCKET_IP_TRANSPARENT,
       ENVOY_SOCKET_IPV6_TRANSPARENT, 1));
   return options;
 }
 
+std::unique_ptr<Socket::Options> SocketOptionFactory::buildLiteralOptions(
+    const Protobuf::RepeatedPtrField<envoy::api::v2::core::SocketOption>& socket_options) {
+  std::unique_ptr<Socket::Options> options = absl::make_unique<Socket::Options>();
+  std::string buf;
+  for (const auto& socket_option : socket_options) {
+    buf.clear();
+    int int_value;
+    switch (socket_option.value_case()) {
+    case envoy::api::v2::core::SocketOption::kIntValue:
+      int_value = socket_option.int_value();
+      buf.append(reinterpret_cast<char*>(&int_value), sizeof(int_value));
+      break;
+    case envoy::api::v2::core::SocketOption::kBufValue:
+      buf.append(socket_option.buf_value());
+      break;
+    default:
+      ENVOY_LOG(warn, "Socket option specified with no value: {}", socket_option.DebugString());
```

Comment:
I think this is just `NOT_REACHED` if you are setting up PGV constraints correctly.

Reply:
What I'm trying to guard against is proto skew in the oneof. I.e. if for some reason we added a new field that wouldn't be covered by the explicit case statements. No value isn't really the right wording though.